src/editor.c

```diff
@@ -1352,6 +1356,150 @@ static gint get_sci_line_code_end_position(ScintillaObject *sci, gint line)
 }
 
 
+/* returns the real indentation width, not the depth */
+static gint get_open_parenthesis_indent(ScintillaObject *sci, gint line, gint *line_matched)
+{
+	gint pos, pos_start = 0;
+	gint ret = -1, depth = 0;
+	gint lexer;
+
+	lexer = sci_get_lexer (sci);
+	pos = sci_get_line_end_position(sci, line) - 1;
+	/*if (one_line)
+		pos_start = sci_get_position_from_line(sci, line);*/
+	for (; pos >= pos_start && depth >= 0; pos--)
+	{
+		if (highlighting_is_code_style (lexer, sci_get_style_at (sci, pos)))
+		{
+			gchar ch = sci_get_char_at(sci, pos);
+
+			if (ch == ')')
+				depth++;
+			else if (ch == '(')
+				depth--;
+		}
+	}
+	if (depth < 0)
+	{
+		gint end_pos;
+		gint match_line;
+
+		end_pos = pos + 1;
+		match_line = sci_get_line_from_position(sci, end_pos);
+		if (line_matched)
+			*line_matched = match_line;
+		/* don't indent if it would use the full line size since it is unlikely
+		 * somebody wants it */
+		if (end_pos < sci_get_line_end_position (sci, match_line) - 1)
+		{
+			ret = sci_get_col_from_position(sci, end_pos + 1);
+			/* and then keep the smallest indent of a line between match and current,
+			 * not to restore a user reduced indent (e.g. for base_indent+1 style) */
+			for (++match_line; match_line < line; ++match_line)
```

Comment:
Well, I find it weird to break alignment if you don't want it for real, and I really find the first example unreadable.
but if it's a common use case, maybe it can be controlled by an option…

I'm not totally sure why I implemented this in the first place (as I doubt I would have actually wrote something needing this willingly myself [1]), but I guess I encountered a situation where I found it was getting in the way of custom indent.

This said, reading the thing a bit more makes me think while the behavior is good in this particular case, the logic is flawed and simply too complex.  What it does here is find the previous unbalanced parenthesis and align to it.  But I guess a more sensible (and actually simpler and cheaper) behavior would be to stop searching whenever we find a balanced line, and use that line's indentation.
This would give this:
```C
  one(1,
    two(2, // manual unindent
        3),
    three(4, // automatically uses L2 indent
          5),
      four(6, // manual reindent
           7),
      five(8, // automatically uses L6 indent
           6)
      );
```
And this also matches better the logic of the other parts.
But with this would remove all code needed to restore manually broken alignment that you seem to want.

---
[1]:  as it's only useful if the parenthesis to match is not the last one on its line (as there is a special case not to create an alignment of the width of the previous line), so something like this doesn't need it:
```c
foo (
  1,
  bar (
    2,
    3
  )
);